### PR TITLE
Issue #2362: Add new 'validateOnlyOverlapping' option for RequireThis

### PR DESCRIPTION
**Pull request by MEZk**
Issue #2362.

### CODE REVIEW

**Comment by mkordas** [comment]
@MEZk thanks for nice PR! Build needs to be fixed though.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java [hunk 1]
```diff
@@ -39,8 +39,12 @@
  * object are explicitly of the form &quot;this.varName&quot; or
  * &quot;this.methodName(args)&quot;.
  * </p>
- *
- * <p>Warning: the Check is very controversial and not that actual nowadays.</p>
+ * Check has the following options:
+ * <p><b>checkFields</b> - whether to check references to fields. Default value is <b>true</b>.</p>
+ * <p><b>checkMethods</b> - whether to check references to methods.
+ * Default value is <b>true</b>.</p>
+ * <p><b>validateOnlyOverlapping</b> - whether to check only overlapping by variables or
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java]: @MEZk sentence needs rewording

**Comment by romani** [comment]
circle-ci is fixed, relaunched.

Travis did awesome work and found exception in Check: 
https://travis-ci.org/checkstyle/checkstyle/jobs/104429420#L517
```
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing /home/travis/build/checkstyle/checkstyle/contribution/checkstyle-tester/src/main/java/spring-framework/spring-webmvc-portlet/src/test/java/org/springframework/web/portlet/mvc/annotation/Portlet20AnnotationControllerTests.java
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:282)
	at org.apache.maven.plugin.checkstyle.exec.DefaultCheckstyleExecutor.executeCheckstyle(DefaultCheckstyleExecutor.java:252)
	at org.apache.maven.plugin.checkstyle.AbstractCheckstyleReport.executeReport(AbstractCheckstyleReport.java:473)
	... 28 more
Caused by: java.lang.ClassCastException: com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck$BlockFrame cannot be cast to com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck$ClassFrame
	at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.canBeReferencedFromStaticContext(RequireThisCheck.java:491)
	at
```

**Comment by MEZk** [comment]
@romani 
Fixed.

**Comment by romani** [comment]
Please activate this Check in checkstyle_checks.xml in separate cimmir

**Comment by romani** [comment]
Some of the changes are not related to your changes, but please do them to ease reading the code
4) 
>
```     private Map<DetailAST, AbstractFrame> frames;
     private AbstractFrame current;
```

please reset these fields in beginTree method , to avoid hard to reproduce problems on multiple files. All non-option fields that keep status of validation need to be reset in beginTree method.

5)
> checkMethod

rename to getMethodWithoutThis(....)

6)
> processField(....)

rename to getFieldWithoutThis(....)

7)
> checkField(....)

rename to smth

**Comment by MEZk** [comment]
@romani 
1) Done.

2) The method has been removed.

3) Reports:
checkstyle - 0 violations
sevntu-checkstyle - 0 violations
[guava](http://mezk.github.io/guava/checkstyle.html)
[findbugs](http://mezk.github.io/findbugs/checkstyle.html)
[pmd](http://mezk.github.io/pmd/checkstyle.html)
[hibernate](http://mezk.github.io/hibernate/checkstyle.html)

4) Done.
5) Done.
6) Done.
7) Done.
8,9,10,11) The check's logic has been changed in the methods.
12) Done.

**Comment by romani** [comment]
Ci detected exceptions, please fix

**Comment by MEZk** [comment]
@romani 
Build failure has been fixed. You can examine all test cases that are covered in InputValidateOnlyOverlappingTrue.java and InputValidateOnlyOverlappingFalse.java

New reports (checkstyle, sevntu-checkstyle, guavam findbugs, pmd, hibernate):
http://mezk.github.io/checkstyle.html

**Comment by romani** [comment]
nice job!

but unfortunately almost all cases in report are false-positives. Cases are not trivial for sure. most of them is related to reassignment of parameter (another Check validation).
Just try to put "this." in places where you have violation and you will damage logic. Please find me online for discussion.

1)
 
> FieldAnnotation.java, MethodAnnotation.java, PackageMemberAnnotation.java, AbstractClassMember.java, ClassFeatureSet.java, .....

argument is reused as local variable, and later used as assignment to a field.
Looks like we need to skip violation if "this." is used in method already for this identifier.

2)

> Churn.java, LockOptions.java

local variable and field used in the same time. We can not force user to use "this." we need to skip such cases.

3)

> PackageStats.java, ByteCodePatternMatch.java

this is pure false-positive.

4)

> TableBinder.java

"this" in static method ?  please extend your UTs Inputs.

5)

> HTMLRenderer.java

false-positive. We can not demand "this." there. It is badly named local variable but it is not related to `Reference to instance variable 'colorize' needs "this.".`

**Comment by MEZk** [comment]
@romani 

**New reports**: 
http://mezk.github.io/RequireThisReport/checkstyle.html
checkstyle - 0 violations
sevntu-checkstyle - 0 violations
guava - 0 violations 
hibernate - 0 violations
spring - 0 violations
orekit - 0 violations
findbugs - 0 violations
pmd - 2 violations
lombook - 0 violations
HBase - 6 violations

5) As it was discussed in Skype:

>
[8:14:41 PM] Андрей Селькин:     
void foo28() {
        boolean booleanField = true;
        booleanField = !booleanField;
}
[8:15:25 PM] RomanIvanov: VIOLATION FOR SURE !

and
>
[8:13:22 PM] RomanIvanov: 
private String stringField;
  String fooR(String stringField ) { 
    stringField = foo(stringField) 
}
[8:13:35 PM] RomanIvanov: VIOLAITON FOR SURE !!!

**Comment by romani** [comment]
merged as FF

Functionality become too complicated, we need release it by parts.
Even I see more rooms for improvement, we will do them later.

**Pull request closed without merging**